lib/libzpool/uzfs_rebuilding.c

```diff
@@ -0,0 +1,296 @@
+	len = metadata_read_chunk_size;
+
+	for (; offset < end; offset += len) {
+		read = 0;
+		len = metadata_read_chunk_size;
+
+		if ((offset + len) > end)
+			len = (end - offset);
+
+		ret = uzfs_read_metadata(snap_zv, buf, offset, len, &read);
+
+		if (read != len || ret)
+			break;
+
+		lun_offset = (offset / metadatasize) * zv->zv_metavolblocksize;
+		for (i = 0; i < len; i += sizeof (blk_metadata_t)) {
+			if (!iszero((blk_metadata_t *)(buf+i)) &&
+			    (compare_blk_metadata((blk_metadata_t *)(buf + i),
+			    low) > 0)) {
+				if (diff_count == 0) {
+					last_lun_offset = lun_offset;
+					last_md = (blk_metadata_t *)(buf+i);
+					last_index = i;
+				}
+				diff_count++;
+				if (last_md != NULL &&
+				    compare_blk_metadata((blk_metadata_t *)
+				    (buf + i), last_md) != 0) {
+					EXECUTE_DIFF_CALLBACK(last_lun_offset,
+					    diff_count, buf, last_index, arg,
+					    last_md, snap_zv, func);
+				}
+			} else if (diff_count) {
+				EXECUTE_DIFF_CALLBACK(last_lun_offset,
+				    diff_count, buf, last_index, arg, last_md,
+				    snap_zv, func);
+			}
+
+			lun_offset += zv->zv_metavolblocksize;
+		}
+
+		if (diff_count) {
+			EXECUTE_DIFF_CALLBACK(last_lun_offset, diff_count, buf,
+			    last_index, arg, last_md, snap_zv, func);
+		}
+	}
+
+	dmu_objset_disown(snap_zv->zv_objset, &snap_zv);
+	umem_free(snap_zv, sizeof (*snap_zv));
+
+	/*
+	 * TODO: if we failed to destroy snapshot here then
+	 * this should be handled separately from application.
+	 */
+	if (end == metaobjectsize)
+		destroy_metadata_snapshot(zv, low);
+
+	umem_free(buf, metadata_read_chunk_size);
+	return (ret);
+}
+
+int
+uzfs_search_nonoverlapping_io(zvol_state_t *zv, uint64_t offset, uint64_t len,
+    blk_metadata_t *metadata, void **list)
+{
+	char *rd_metadata_buf;
+	uint64_t rd_rlen;
+	metaobj_blk_offset_t rd_metablk;
+	blk_metadata_t *rd_metadata;
+	int diff_count = 0;
+	int count = 0;
+	int ret = 0;
+	int i = 0;
+	uint64_t lun_offset = 0, last_lun_offset = 0;
+	list_t *chunk_list = NULL;
+	uint64_t metavolblocksize = zv->zv_metavolblocksize;
+	uint64_t metadatasize = zv->zv_volmetadatasize;
+
+	get_zv_metaobj_block_details(&rd_metablk, zv, offset, len);
+	rd_metadata_buf = umem_alloc(rd_metablk.m_len, UMEM_NOFAIL);
+
+	ret = uzfs_read_metadata(zv, rd_metadata_buf, rd_metablk.m_offset,
+	    rd_metablk.m_len, &rd_rlen);
+	if (ret || rd_rlen != rd_metablk.m_len) {
+		printf("failed to read metadata\n");
+		goto exit;
+	}
+
+	chunk_list = umem_alloc(sizeof (*chunk_list), UMEM_NOFAIL);
+	list_create(chunk_list, sizeof (uzfs_io_chunk_list_t),
+	    offsetof(uzfs_io_chunk_list_t, link));
+
+	for (i = 0; i < rd_metablk.m_len; i += sizeof (blk_metadata_t)) {
+		rd_metadata = (blk_metadata_t *)(rd_metadata_buf + i);
+		lun_offset = ((rd_metablk.m_offset + i) * metavolblocksize) /
+		    metadatasize;
+		ret = compare_blk_metadata(rd_metadata, metadata);
+		if (ret == -1) {
+			// old io number < new io number
```

Comment:
on_disk io number < incoming IO number

Reply:
Ok. that makes sense. I will update comment.